ws/client.go

```diff
@@ -106,94 +261,49 @@ func (c *Client) readPump() {
 		}
 		message = bytes.TrimSpace(bytes.Replace(message, newline, space, -1))
 
-		var r req
-		if err = json.Unmarshal(message, &r); err != nil {
-			c.send <- []byte(err.Error())
-			continue
-		}
-
-		ctx, cancel := context.WithTimeout(context.Background(), messageTimeout)
-		switch r.Type {
-		case "add-thread":
-			var m addThreadMsg
-			if err = json.Unmarshal([]byte(r.Msg), &m); err != nil {
-				c.send <- []byte(err.Error())
-				continue
-			}
-
-			var info thread.Info
-			if m.Addr != "" {
-				if m.FollowKey == "" {
-					c.send <- []byte("follow key is required with address")
-					continue
-				}
-
-				fk, err := parseKey(m.FollowKey)
-				if err != nil {
-					c.send <- []byte(err.Error())
-					continue
-				}
-				var rk *sym.Key
-				if m.ReadKey != "" {
-					rk, err = parseKey(m.ReadKey)
-					if err != nil {
-						c.send <- []byte(err.Error())
-						continue
-					}
-				}
-				maddr, err := ma.NewMultiaddr(m.Addr)
-				if err != nil {
-					c.send <- []byte(err.Error())
-					continue
-				}
-
-				info, err = c.hub.service.AddThread(ctx, maddr, fk, rk)
-				if err != nil {
-					c.send <- []byte(err.Error())
-					continue
-				}
-			} else {
-				info, err = util.CreateThread(c.hub.service, thread.NewIDV1(thread.Raw, 32))
-				if err != nil {
-					c.send <- []byte(err.Error())
-					continue
-				}
-			}
-
-			log.Debugf("added thread %s", info.ID)
-
-		case "pull-thread":
-			c.send <- []byte("todo")
-		case "delete-thread":
-			c.send <- []byte("todo")
-		case "add-follower":
-			c.send <- []byte("todo")
-		case "add-record":
-			c.send <- []byte("todo")
-		case "get-record":
-			c.send <- []byte("todo")
-		case "subscribe":
-			var m subscribeMsg
-			if err = json.Unmarshal([]byte(r.Msg), &m); err != nil {
-				c.send <- []byte(err.Error())
-				continue
+		callerID := "unknown"
+		var caller rpcCaller
+		var result interface{}
+		if err = json.Unmarshal(message, &caller); err == nil {
+			callerID = caller.ID
+
+			ctx, cancel := context.WithTimeout(context.Background(), rpcCallTimeout)
+			switch caller.Method {
+			case "addThread":
+				result, err = c.addThread(ctx, caller.Args...)
+			case "pullThread":
+				err = fmt.Errorf("todo")
+			case "deleteThread":
+				err = fmt.Errorf("todo")
+			case "addFollower":
+				err = fmt.Errorf("todo")
+			case "addRecord":
```

Comment:
Why not these ones yet? Specific reason, or just not there yet?

Reply:
Oops! I forgot to enable the new methods... added.